### PR TITLE
Carousel: Make sure we show the caption

### LINKED ISSUE #20295: Carousel: caption is not displayed in modal anymore
In Jetpack 9.8.1, when using the Carousel feature, the Carousel view displayed 3 different types of information:
- The image title
- The image description
- The image's description

image

In Jetpack 9.9, following the changes in #20107, only 2 types of information are displayed:

- The image title
- The image description

Screenshot 2021-07-07 at 15 29 58

That's problematic for the folks who were previously using captions to display important information. Since we're also displaying the same info twice, we should be able to revert to the previous behaviour.

********

Internal references:
- p1625660123167900-slack-C0CMN0V97
- 4122177-zen

### PR DESCRIPTION
**Pull request by apeatling**
Fixes https://github.com/Automattic/jetpack/issues/20295

#### Changes proposed in this Pull Request:

* Show the caption in the carousel in the bottom bar, and in the more info.

| Before        | After           |
| ------------- |:-------------:| 
| Screen Shot 2021-07-07 at 9 15 54 AM  | Screen Shot 2021-07-07 at 9 13 09 AM |

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Open a carousel with and image that has a caption.
* Confirm that you see the caption in the bottom bar, and the info section.

### CODE REVIEW

File: projects/plugins/jetpack/modules/carousel/jetpack-carousel.js
Comment (stacimc) [inline projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:740]: If I'm understanding correctly, this is trying to hide the caption underneath the photo if its text is identical to the title? It looks like previously we hid the title in the extended info section instead of the caption. Nit -- if we keep this, it would be clearer to name this variable `caption`.

I tested this by setting the title, caption, and description to identical text. The description was correctly hidden, but the title and caption both displayed in the extended info section, and the caption under the photo displayed as well 🤔 I wonder if it would make more sense to hide the `.jp-carousel-caption` (the one in the extended info section) when the title and caption match?

**Review by stacimc — CHANGES_REQUESTED** [review]
This is working well for me with title, caption, and description specified 👍 

I noticed that the captions only display in the expanded info view on mobile, so no caption underneath the photo (is that the intended behavior?)

Also left a comment about what happens when the title/caption/description are the same, and it looks like there are just a few linting errors.


Worth noting: I managed to get into a very odd state where none of the three displayed when I:
1. Added an image to the gallery
2. Used the Media Library to add the same image again, and edited its title/caption at the same time

That appears to be an issue with attributes getting set/unrelated to this PR. I'll test some more and make a separate issue if necessary.

**Comment by jeherve** [comment]
> Worth noting: I managed to get into a very odd state where none of the three displayed when I:
> 
> 1. Added an image to the gallery
> 2. Used the Media Library to add the same image again, and edited its title/caption at the same time
> 
> That appears to be an issue with attributes getting set/unrelated to this PR. I'll test some more and make a separate issue if necessary.

Yes, I believe you ran into the annoying problem we were discussing in #11594.

**Comment by apeatling** [comment]
> I noticed that the captions only display in the expanded info view on mobile, so no caption underneath the photo (is that the intended behavior?)

Yes, there simply is not enough space once the viewport width becomes narrow. I don't think we should be overlaying the captions on the image here because it's about viewing the image.

I've updated the caption to fill more of the bottom bar space when it is visible.

File: projects/plugins/jetpack/modules/carousel/jetpack-carousel.js
Comment (apeatling) [reply projects/plugins/jetpack/modules/carousel/jetpack-carousel.js]: I think I've fixed this now. Good catch. I've completely removed this bit of code since it was conflicting with the function that shows/hides the other data. I think it's ready for another look.

File: projects/plugins/jetpack/modules/carousel/jetpack-carousel.css
Comment (stacimc) [inline projects/plugins/jetpack/modules/carousel/jetpack-carousel.css:342]: By no means a blocking issue, but I realized it's quite confusing that this `.jp-carousel-photo-title-container` class is used on what's actually the _caption_ underneath the photo. We could probably clean that up in a follow up.

**Review by stacimc — DISMISSED** [review]
Tested setting title/caption/description to identical values and LGTM!

I'll leave this up to you and I think this is good to go, but in my opinion when the caption and title match it might look nicer to hide the caption in the info section (leaving the caption under the photo), rather than hiding the title. I think it looks a bit odd in the info section with just the smaller caption text rather than the larger title:

With title hidden:
Screen Shot 2021-07-08 at 11 07 36 AM

With caption hidden:
Screen Shot 2021-07-08 at 11 08 59 AM

**Comment by antermoia** [comment]
I can't see the full, untruncated caption in the "view more/comments" section UNLESS I have the "show EXIF" option enabled (which I'd rather leave disabled).

**Comment by apeatling** [comment]
@antermoia I think we can fix this for you in a follow up PR.

File: projects/plugins/jetpack/modules/carousel/jetpack-carousel.php
Comment (jeherve) [inline projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:692]: Should we check that `get_post` isn't null before we start pulling data from `$attachment`?

You'll also need to address the indenting issues so we can merge this.

**Review by jeherve — CHANGES_REQUESTED** [review]

**Comment by antermoia** [comment]
It would also appear that in "New Carousel", the caption is now completely sanitized, and all HTML is stripped from it. I have a lot of captions with a link to an external site in them ([example](https://www.alex-kunz.com/portfolio/2020-flower-plant-portfolio/)) I take it that's not possible anymore?

File: projects/plugins/jetpack/modules/carousel/jetpack-carousel.php
Comment (apeatling) [reply projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:692]: This is old code, I'll add a check here though.

**Comment by apeatling** [comment]
> all HTML is stripped 

I've removed this now.

**Review by jeherve — DISMISSED** [review]
I spotted an issue, and would have a question:

- There seem to be encoding issues:

**Stable**

Screenshot 2021-07-13 at 13 57 09

**With this branch**

Screenshot 2021-07-13 at 13 56 48

- Would it be possible to display the title when no caption is available, instead of having an empty field?

**Stable**

Screenshot 2021-07-13 at 13 59 37

**With this branch**

Screenshot 2021-07-13 at 13 59 24

**Comment by apeatling** [comment]
@jeherve I can't seem to reproduce this encoding issue at all. I'm adding a curly apostrophie but it's coming out correctly, even if I insert the html entity, and adding it in master and switching branches:

Screen Shot 2021-07-13 at 12 17 04 PM

**Comment by apeatling** [comment]
> Would it be possible to display the title when no caption is available, instead of having an empty field?

Fixed 👍

**Comment by sdixon194** [comment]
Hmm, I think maybe I just need clarity on how captions work. On my test site:

- One photo (the building with the caption "test") - doesn't have any caption at all, even though one is specified in the editor (and without it, the title doesn't display as well): https://linear-peccary.jurassic.ninja/2021/07/13/gallery/ No console error anything.
- Same here, can't figure out why the desk photo doesn't show a caption at all: https://linear-peccary.jurassic.ninja/2021/07/13/gallery-three/
- On this post, I added a caption for the bridge in the media library, but changed it in the editor. However the carousel caption is the media library copy, instead of what I'd expect (the one I added in the editor). I think this is expected behavior? https://linear-peccary.jurassic.ninja/2021/07/13/gallery-two/

Happy to DM credentials if you want to play around!

**Review by sdixon194 — APPROVED** [review]
Everything works well for me now, outstanding issues have been fixed.

**Comment by sdixon194** [comment]
Cherry picked to release branch in fc3bc9b7729e07d74ab431fe2fadd3db46dcb379

**Pull request merged**
